### ISSUE TITLE
Asking Help for not Getting the Promising Results

### ISSUE THREAD

**Issue by Yingminzhou** [cheungdaven/metricfactorization#2]
@cheungdaven 
hi, thanks for providing the source code. 
After reading it, I have few questions:
- Does the parameter [beta] (appeared in line 39 of [MetricFRanking](https://github.com/cheungdaven/metricfactorization/blob/master/MetricFRanking.py)) reference to the distance scale [a] in the paper?
- I ran the ranking code on the [dataset](https://github.com/cheungdaven/metricfactorization/blob/master/data/filmtrust_ratings.dat) using the following parameters:
	- alpha: 4 (confidence level, as referred in the paper)
	- learning_rate: 0.05 (as referred in the paper)
	- epoch: 200
	- N: 100 (latent factor dimension)
	- batch_size: 1024 * 3
	- beta: 2.25 (distance scale, as referred in the paper)

But I found that the loss seems to be non-convergent, and the metrics in terms of precision/recall/map/mrr/ndcg are not so promising. Could you please provide some advice for trouble-shotting this problem? 

 Thanks for your kind help in advance.

**Comment by cheungdaven** [comment]
> @cheungdaven
> hi, thanks for providing the source code.
> After reading it, I have few questions:
> 
> * Does the parameter [beta] (appeared in line 39 of [MetricFRanking](https://github.com/cheungdaven/metricfactorization/blob/master/MetricFRanking.py)) reference to the distance scale [a] in the paper?
> * I ran the ranking code on the [dataset](https://github.com/cheungdaven/metricfactorization/blob/master/data/filmtrust_ratings.dat) using the following parameters:
>   
>   * alpha: 4 (confidence level, as referred in the paper)
>   * learning_rate: 0.05 (as referred in the paper)
>   * epoch: 200
>   * N: 100 (latent factor dimension)
>   * batch_size: 1024 * 3
>   * beta: 2.25 (distance scale, as referred in the paper)
> 
> But I found that the loss seems to be non-convergent, and the metrics in terms of precision/recall/map/mrr/ndcg are not so promising. Could you please provide some advice for trouble-shotting this problem?
> 
> Thanks for your kind help in advance.

I have updated the code, u can run it again. Note that with the former code, u may be not able to get valid results on GPU.

**Comment by Yingminzhou** [comment]
@cheungdaven 
Thanks for your reply and update. It seems that `self.optimizer` is not initialized in the newest code. I set it to the `gds` directly. After running this code, we got the following results, and the loss varies unexpectedly ranging from 17.0 to 24.0.

- precision@10: 0.10
- recall@10: 0.34
- precision@5: 0.21
- recall@5: 0.14
- map: 0.23
- mrr:  0.57
- ndcg: 0.70

I still have a question: does the loss come to convergence on the given data set in your experiment? and how's the results in your experiment?

Thanks very much!

**Comment by cheungdaven** [comment]
> @cheungdaven
> Thanks for your reply and update. It seems that `self.optimizer` is not initialized in the newest code. I set it to the `gds` directly. After running this code, we got the following results, and the loss varies unexpectedly ranging from 17.0 to 24.0.
> 
> * precision@10: 0.10
> * recall@10: 0.34
> * precision@5: 0.21
> * recall@5: 0.14
> * map: 0.23
> * mrr:  0.57
> * ndcg: 0.70
> 
> I still have a question: does the loss come to convergence on the given data set in your experiment? and how's the results in your experiment?
> 
> Thanks very much!


Sorry for the mistakes. I updated it again. The loss actually does not matter that much here. I tested the model again without changing anything (on CPU), this is what I get in the first two epochs:
Index: 0001; Epoch: 0000; loss = 98.394317627
Index: 1001; Epoch: 0000; loss = 46.941368103
precision@10:0.3414364640883978
recall@10:0.6124396180159487
precision@5:0.409629044988161
recall@5:0.3932369676444957
map:0.4774422598338411
mrr:0.6045688482104703
ndcg:0.6420997034191704
Index: 0001; Epoch: 0001; loss = 31.360254288
Index: 1001; Epoch: 0001; loss = 28.493558884
Index: 0001; Epoch: 0002; loss = 20.149700165
Index: 1001; Epoch: 0002; loss = 21.184005737
precision@10:0.3572217837411208
recall@10:0.6504516867604051
precision@5:0.4369376479873718
recall@5:0.44170152517740147
map:0.5263272663303024
mrr:0.6642867287042593
ndcg:0.6826968291038207

The performance will be better if it continued. I guess something went wrong with your environment, r u using python 3?

**Comment by Yingminzhou** [comment]
@cheungdaven Thanks for your reply.
The results seem very exciting! I am running in the following environment on gpu:
- python: 2.7.5
- tensorflow: 1.1.0

Will these settings affect the results?

Thanks.

**Comment by cheungdaven** [comment]
> @cheungdaven Thanks for your reply.
> The results seem very exciting! I am running in the following environment on gpu:
> 
> * python: 2.7.5
> * tensorflow: 1.1.0
> 
> Will these settings affect the results?
> 
> Thanks.

The code may suffer from precision issues with python 2.7, u can change to python 3+ to see what happens.

**Comment by Yingminzhou** [comment]
@cheungdaven 
Awesome! After changing the python version to 3.6, we also got the exciting results!

By the way, if we want to apply this wonderful model to a large-scale dataset (such as 2 kw users and 2w items), do you have any good suggestions?

Thanks so much!

**Comment by cheungdaven** [comment]
@Yingminzhou 
A straightforward way is to use multiple GPUs.
Negative sampling might influence the performance.

**Comment by Yingminzhou** [comment]
@cheungdaven  
Thanks for your reply. 
We came cross the memory overload problem while loading the  data (1500w users 2w items and 4500w ratings). Maybe we could try to optimize the memory storage in the future.

Thanks a lot for your kind help!

**Comment by cheungdaven** [comment]
@Yingminzhou 
You are welcome. Yeah, the data load function can be optimized lol. I will include this project into the DeepRec (https://github.com/cheungdaven/DeepRec), and will keep updating that one if I have time. 

Thank you for ur interests.
